### PR TITLE
Fix issue 12805: @nogc std.range.iota(FP)

### PR DESCRIPTION
**Pull request by JackStouffer**

### CODE REVIEW

**Comment by MetaLang** [comment]
Why is @nogc not inferred for `to` when the source type is `int` and the target type is `size_t`?

**Comment by JackStouffer** [comment]
>Why is @nogc not inferred for to when the source type is int and the target type is size_t?

Not really something I have the answer to.

**Comment by JakobOvrum** [comment]
> Why is `@nogc` not inferred for `to` when the source type is `int` and the target type is `size_t`?

Probably because `to` does range checking; it allocates and throws an exception if the input doesn't fit in the target type.

It's the general issue with exceptions. I don't think removing this correctness feature from `iota` is a good workaround.

**Comment by MetaLang** [comment]
Didn't @MartinNowak recently submit a pull request for a `staticError` helper? Can we reuse that here?

**Comment by JakobOvrum** [comment]
It's for `Error`s, under the assumption that they won't chain.

File: std/range/package.d [hunk 1]
```diff
@@ -4710,12 +4710,10 @@ body
 
         this(Value start, Value end, Value step)
         {
-            import std.conv : to;
-
             this.start = start;
             this.step = step;
             immutable fcount = (end - start) / step;
-            count = to!size_t(fcount);
+            count = cast(size_t) fcount;
```
Comment (DmitryOlshansky) [inline std/range/package.d]: Narrowing conversion

**Comment by JackStouffer** [comment]
What if I put this line either in the contract or before the cast?

```
assert((end - begin) / step < size_t.max, "iota: calling with these values will result in an overflow");
```

**Comment by quickfur** [comment]
I don't think it's a good idea to change `to!size_t` to `cast(size_t)`. The former does range-checking, the latter does not.

The correct fix is to address the issue of exceptions in `@nogc` code, which is an ongoing issue that not only affects this PR, but also many other situations that people have been complaining about.

**Comment by DmitryOlshansky** [comment]
> The correct fix is to address the issue of exceptions in @nogc code, which is an ongoing issue that not only affects this PR, but also many other situations that people have been complaining about.

Until Exceptions are ref-counted there is not much to do aside from avoiding them in the first place. Certianly we can't silently change behaviour of functions that are throwing exceptions.

**Comment by MetaLang** [comment]
Can't the except be statically allocated?

**Comment by JackStouffer** [comment]
While I agree that the correct fix is to make exceptions @nogc, but that could be years from now. In the mean time, how about

```
debug
{
    count = to!size_t(fcount);
}
count = cast(size_t) fcount;
```

**Comment by schveiguy** [comment]
What we need is a `@nogc` version of `to`, that may take an additional error parameter to indicate a problem. I don't think the solution is to avoid the gc for exceptions here.

**Comment by MetaLang** [comment]
I remember Bearophile suggesting quite awhile ago that we have a version of `to` that catches the exception and returns a `Nullable!Target`. That seems like a worthwhile pursuit to me, as much as I hate the current implementation of `Nullable`.

**Comment by schveiguy** [comment]
> a version of to that catches the exception

This makes it `nothrow`, but not `@nogc`

A possibility is to have `to` accept a callable to call if the conversion fails, which could then in the throwing version, just throw an exception. Then at least the `to` implementation is separate from the error handling.

**Comment by quickfur** [comment]
Hmm. So something along these lines?
````
auto to(S,T)(S src, T target, lazy Exception e = new Exception(...)) { ... }
````

**Comment by MetaLang** [comment]
Wut? Catching an exception forfeits @nogc? Something seems wrong there... Why does catching an exception need to allocate?

**Comment by MetaLang** [comment]
@quickfur Nice idea, didn't think of that.

**Comment by quickfur** [comment]
No, catching an exception doesn't forfeit `@nogc`. The problem is that by the time you catch the exception, it has already been allocated (i.e., no longer `@nogc`).

**Comment by schveiguy** [comment]
> So something along these lines

I'm not sure that will work. As a private implementation function it should be fine, but as a general call, I think `to` should decide the message for the exception. I was thinking more like this:

```D
T to(Handler, S, T)(S src, scope Handler h)
{
   ...
   if(error)
   {
      h(errmsg);
      return T.init;
   }
   ...
}
```

**Comment by quickfur** [comment]
This may address this particular case, but what about exceptions in Phobos in general? With the current push for `@nogc`, I think we should think about a more generic solution that can be used across the board for Phobos.  Otherwise we're going to have to invent ad hoc workarounds every time somebody comes across something else in Phobos that should be `@nogc` but unfortunately throws exceptions. (Which makes exceptions kinda a joke in Phobos, they're part of the language but we effectively can't use them for fear of breaking `@nogc`.)

**Comment by jpf91** [comment]
@quickfur The solution is to somehow integrate memory management into the exception system. Right now exceptions must use some kind of automated memory management (GC, static allocation) and it must work with classes derived from Exception. IIRC Andrei's long term plan is to make Exceptions reference counted. Another solution is to keep a reference to the allocator in the Exception and provide a dispose() method. This method should then be called either manually or automatically by the compiler in catch blocks. It gets a little bit more complicated because of Exception chaining.

**Comment by JackStouffer** [comment]
I'm going to go ahead and close this, as the consensus seems to be that correct solution is to make exceptions reference counted, which will make this pull null and void.

**Pull request closed without merging**